sound/soc/sof/intel/hda-dsp.c

```diff
@@ -282,6 +282,52 @@ void hda_dsp_ipc_int_disable(struct snd_sof_dev *sdev)
 			HDA_DSP_REG_HIPCCTL_BUSY | HDA_DSP_REG_HIPCCTL_DONE, 0);
 }
 
+static void hda_dsp_wait_d0i3c_done(struct snd_sof_dev *sdev, int *retry)
+{
+	struct hdac_bus *bus = sof_to_bus(sdev);
+	u8 reg;
+
+	reg = snd_hdac_chip_readb(bus, VS_D0I3C);
+	while ((reg & SOF_HDA_VS_D0I3C_CIP) && --*retry) {
+		usleep_range(10, 15);
+		reg = snd_hdac_chip_readb(bus, VS_D0I3C);
+	}
+}
+
+int hda_dsp_set_state(struct snd_sof_dev *sdev,
+		      enum sof_d0_substate d0_substate)
+{
+	struct hdac_bus *bus = sof_to_bus(sdev);
+	u8 value;
+	int retry = 50;
+
+	/* Write to D0I3C after Command-In-Progress bit is cleared */
+	hda_dsp_wait_d0i3c_done(sdev, &retry);
```

Comment:
@keyonjie why do we need to pass this as a pointer? WHy not just initialize the  check the return value for hda_dsp_wait_d0i3c_done(). It seems a bit of an overkill to pass a pointer for no good reason

Reply:
@keyonjie do we need to add a check to see if the previous substate is already the requested state and return in tthat case?

Reply:
agree, remove the parameter and return -ETIMEDOUT if retries exhausted, maybe also use a macro for the retry count